pkg/storage/etcd/etcd_watcher.go

```diff
@@ -390,8 +413,9 @@ func (w *etcdWatcher) Stop() {
 	w.stopLock.Lock()
 	defer w.stopLock.Unlock()
 	// Prevent double channel closes.
-	if !w.stopped {
-		w.stopped = true
-		close(w.userStop)
+	if w.cancel != nil {
+		w.cancel()
+		w.cancel = nil
 	}
+	close(w.userStop)
```

Comment:
But we can potentially still close the channel more than once now. It doesn't seem to be safe.